File: sdk/go/provider/mailmomy.go
package provider

import (
	"encoding/json"
	"fmt"
	"io"
	"math/rand"
	"net/url"

	http "github.com/bogdanfinn/fhttp"
)

// mailmomy.com：免注册、无鉴权、纯 GET JSON API 临时邮箱。
// token = email 无状态模式；MX 自研 mail.mailmomy.com。

const (
	mailmomyBaseURL = "https://mailmomy.com"
	// 规格指定的固定 UA，与 npm 端一致
	mailmomyUA = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/149.0.0.0 Safari/537.36"
)

func mailmomyRandomLocal(length int) string {
	const chars = "abcdefghijklmnopqrstuvwxyz0123456789"
	buf := make([]byte, length)
	for i := range buf {
		buf[i] = chars[rand.Intn(len(chars))]
	}
	return string(buf)
}

// mailmomyPickDomain 拉取当前可用域名池并随机选取。
// GET /api/domains/active → JSON 字符串数组；请求失败或列表为空时回退 mailmomy.com。
func mailmomyPickDomain() string {
	const fallback = "mailmomy.com"
	client := HTTPClient()
	req, err := http.NewRequest("GET", mailmomyBaseURL+"/api/domains/active", nil)
	if err != nil {
		return fallback
	}
	req.Header.Set("User-Agent", mailmomyUA)
	req.Header.Set("Accept", "application/json")

	resp, err := client.Do(req)
	if err != nil {
		return fallback
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		return fallback
	}

	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return fallback
	}

	var list []string
	if err := json.Unmarshal(raw, &list); err != nil {
		return fallback
	}
	usable := make([]string, 0, len(list))
	for _, d := range list {
		if d != "" {
			usable = append(usable, d)
		}
	}
	if len(usable) == 0 {
		return fallback
	}
	return usable[rand.Intn(len(usable))]
}

// MailmomyGenerate 创建 mailmomy.com 临时邮箱
func MailmomyGenerate() (*CreatedMailbox, error) {
	domain := mailmomyPickDomain()
	email := fmt.Sprintf("%s@%s", mailmomyRandomLocal(10), domain)
	return &CreatedMailbox{
		Channel: "mailmomy",
		Email:   email,
		Token:   email,
	}, nil
}

// mailmomyMessage mailmomy 邮件消息
type mailmomyMessage struct {
	ID         string `json:"id"`
	Recipient  string `json:"recipient"`
	From       string `json:"from"`
	Subject    string `json:"subject"`
	Message    string `json:"message"`
	BodyText   string `json:"bodyText"`
	ReceivedAt string `json:"receivedAt"`
}

// MailmomyGetEmails 获取 mailmomy.com 邮件列表
// GET /api/mail/messages?to=<email>&page=1&limit=20
// 返回 {emails:[{id,recipient,from,subject,message,bodyText,receivedAt}], total, page, limit, pages}
func MailmomyGetEmails(email string) ([]NormEmail, error) {
	if email == "" {
		return nil, fmt.Errorf("mailmomy: 缺少邮箱地址")
	}

	client := HTTPClient()
	u := fmt.Sprintf("%s/api/mail/messages?to=%s&page=1&limit=20", mailmomyBaseURL, url.QueryEscape(email))
	req, err := http.NewRequest("GET", u, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("User-Agent", mailmomyUA)
	req.Header.Set("Accept", "application/json")

	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if err := CheckHTTPStatus(resp, "mailmomy getEmails"); err != nil {
		return nil, err
	}

	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var data struct {
		Emails []mailmomyMessage `json:"emails"`
	}
	if err := json.Unmarshal(raw, &data); err != nil {
		return nil, err
	}
	if len(data.Emails) == 0 {
		return []NormEmail{}, nil
	}

	emails := make([]NormEmail, 0, len(data.Emails))
	for _, msg := range data.Emails {
		to := msg.Recipient
		if to == "" {
			to = email
		}
		text := msg.BodyText
		if text == "" {
			text = msg.Message
		}
		flat := map[string]interface{}{
			"id":      msg.ID,
			"from":    msg.From,
			"to":      to,
			"subject": msg.Subject,
			"text":    text,
			"html":    msg.Message,
			"date":    msg.ReceivedAt,
			"isRead":  false,
		}
		emails = append(emails, NormalizeMap(flat, email))
	}
	return emails, nil
}

File: sdk/npm/src/providers/mailmomy.ts
import { InternalEmailInfo, Email, Channel } from "../types";
import { normalizeEmail } from "../normalize";
import { fetchWithTimeout } from "../retry";

const CHANNEL: Channel = "mailmomy";
const BASE_URL = "https://mailmomy.com";

const UA =
  "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/149.0.0.0 Safari/537.36";

function randomLocal(n: number): string {
  const chars = "abcdefghijklmnopqrstuvwxyz0123456789";
  let r = "";
  for (let i = 0; i < n; i++) {
    r += chars[Math.floor(Math.random() * chars.length)];
  }
  return r;
}

/**
 * 拉取 mailmomy 当前可用域名池。
 * GET /api/domains/active → JSON string array
 * 优先选包含 "mailmomy" 或 ".com" 的域名（较稳定），全池均可用。
 */
async function pickDomain(): Promise<string> {
  const resp = await fetchWithTimeout(`${BASE_URL}/api/domains/active`, {
    method: "GET",
    headers: { "User-Agent": UA, Accept: "application/json" },
  });
  if (!resp.ok) {
    /* 域名接口失败时回退到默认 mailmomy.com */
    return "mailmomy.com";
  }
  const data = (await resp.json()) as unknown;
  if (!Array.isArray(data) || data.length === 0) {
    return "mailmomy.com";
  }
  const list = data.filter(
    (d): d is string => typeof d === "string" && d.length > 0,
  );
  if (list.length === 0) return "mailmomy.com";
  return list[Math.floor(Math.random() * list.length)];
}

/**
 * 创建 mailmomy.com 临时邮箱
 * 服务免注册，直接构造 <local>@<域名> 即可收信
 */
export async function generateEmail(): Promise<InternalEmailInfo> {
  const domain = await pickDomain();
  const email = `${randomLocal(10)}@${domain}`;
  return {
    channel: CHANNEL,
    email,
    token: email,
  };
}

/**
 * 获取 mailmomy.com 邮件列表
 * GET /api/mail/messages?to=<email>&page=1&limit=20
 * 返回 {emails:[{id,recipient,from,subject,message,bodyText,receivedAt}], total, page, limit, pages}
 */
export async function getEmails(
  token: string,
  email: string,
): Promise<Email[]> {
  const addr = token || email;
  if (!addr) {
    throw new Error("mailmomy: 缺少邮箱地址");
  }

  const url = `${BASE_URL}/api/mail/messages?to=${encodeURIComponent(addr)}&page=1&limit=20`;
  const resp = await fetchWithTimeout(url, {
    method: "GET",
    headers: { "User-Agent": UA, Accept: "application/json" },
  });

  if (!resp.ok) {
    throw new Error(`mailmomy: 获取邮件失败 ${resp.status}`);
  }

  const data = (await resp.json()) as MailmomyListResponse;
  if (!Array.isArray(data.emails)) {
    return [];
  }

  return data.emails.map((msg) =>
    normalizeEmail(
      {
        id: String(msg.id ?? ""),
        from: msg.from || "",
        to: msg.recipient || addr,
        subject: msg.subject || "",
        text: msg.bodyText || msg.message || "",
        html: msg.message || "",
        date: msg.receivedAt || "",
        isRead: false,
        attachments: [],
      },
      addr,
    ),
  );
}

type MailmomyMessage = {
  id?: string;
  recipient?: string;
  from?: string;
  subject?: string;
  message?: string;
  bodyText?: string;
  receivedAt?: string;
};

type MailmomyListResponse = {
  emails?: MailmomyMessage[];
  total?: number;
  page?: number;
  limit?: number;
  pages?: number;
};
